ops/pebble.py

```diff
@@ -0,0 +1,537 @@
+            id=ChangeID(d['id']),
+            kind=d['kind'],
+            summary=d['summary'],
+            status=d['status'],
+            tasks=[Task.from_dict(t) for t in d.get('tasks') or []],
+            ready=d['ready'],
+            err=d.get('err'),
+            spawn_time=_parse_timestamp(d['spawn-time']),
+            ready_time=_parse_timestamp(d['ready-time']) if d.get('ready-time') else None,
+        )
+
+    def __repr__(self):
+        tasks_str = ',\n'.join(textwrap.indent(repr(t), '    ') for t in self.tasks or [])
+        return """Change(
+    id={self.id!r},
+    kind={self.kind!r},
+    summary={self.summary!r},
+    status={self.status!r},
+    tasks={tasks},
+    ready={self.ready!r},
+    err={self.err!r},
+    spawn_time={self.spawn_time!r},
+    ready_time={self.ready_time!r},
+)""".format(
+            self=self,
+            tasks='[\n' + textwrap.indent(tasks_str, '    ') + ',\n    ]' if self.tasks else '[]',
+        )
+
+    __str__ = __repr__
+
+
+class API:
+    """Pebble API client."""
+
+    def __init__(self, socket_path=None, opener=None, base_url='http://localhost', timeout=5.0):
+        """Initialize an instance. Defaults to Unix socket "$PEBBLE/.pebble.socket"."""
+        if opener is None:
+            opener = self._get_default_opener(socket_path)
+        self.opener = opener
+        self.base_url = base_url
+        self.timeout = timeout
+
+    @classmethod
+    def _get_default_opener(cls, socket_path):
+        if socket_path is None:
+            PEBBLE = os.getenv('PEBBLE')
+            if not PEBBLE:
+                raise Exception('You must specify socket_path or set $PEBBLE')
+            socket_path = os.path.join(PEBBLE, '.pebble.socket')
+
+        opener = urllib.request.OpenerDirector()
+        opener.add_handler(_UnixSocketHandler(socket_path))
+        opener.add_handler(urllib.request.HTTPDefaultErrorHandler())
+        opener.add_handler(urllib.request.HTTPRedirectHandler())
+        opener.add_handler(urllib.request.HTTPErrorProcessor())
+        return opener
+
+    def _request(self, method: str, path: str, query: Dict = None, body: Dict = None) -> Dict:
+        url = self.base_url + path
+        if query:
+            url = url + '?' + urllib.parse.urlencode(query)
+
+        headers = {'Accept': 'application/json'}
+        data = None
+        if body is not None:
+            data = json.dumps(body).encode('utf-8')
+            headers['Content-Type'] = 'application/json'
+
+        request = urllib.request.Request(url, method=method, data=data, headers=headers)
+        response = self.opener.open(request, timeout=self.timeout)
+        result = json.load(response)
+        return result
+
+    def get_system_info(self) -> SystemInfo:
+        """Get system info."""
+        result = self._request('GET', '/v1/system-info')
+        return SystemInfo.from_dict(result['result'])
+
+    def get_warnings(self, select=WarningState.PENDING) -> List[Warning]:
```

Comment:
You need to type hint here that `select` is `WarningState`, and that it's optional, right? Or that is done automatically using the provided default?

Reply:
Good call -- fixed. No, it's not done automatically.